man/vde_over_ns.1
.TH VDE_OVER_NS 1 "May 2, 2007" "Virtual Distributed Ethernet"
.SH NAME
vde_over_ns \- Steganographic cable over NS protocol for Virtual Distributed Ethernet
.SH SYNOPSIS
.B vde_over_ns
[\fB\-D\fR]
[\fB\-c\fI server_address\fR]
[\fB\-s\fI vde_plug\fR]
[\fB\-i\fI ip.to.bi.nd\fR]
\fIdomain.name

.br
.SH DESCRIPTION
A
\fBvde_over_ns\fP 
allows one to connect two remote vde_switches through steganographic NS query/response datagrams.
e.g.:

.B
dpipe vde_plug /tmp/s1 = vde_over_ns tun.mydomain.vde 

binds to udp port 53 and listen for incoming steganographic NS queries (server mode).
All the traffic to/from stdout/stdin is redirected to the client via ns responses.

creates a wire between two vde_switches (with sockets /tmp/s1 and /tmp/s2
respectively). This cable looses 10% of the packets in each direction.

The same vde_over_dns server can be created using:

.B
vde_over_ns -s /tmp/s1 tun.mydomain.vde

The syntax is almost identical for the client mode, except that
to have a vde_over_ns client connect to that server, the \fB -c \fIserver\fR 
option must be specified.

.SH OPTIONS
.TP
.B \-s "\fIvdeplug\fP"
If specified, the vde_over_ns will connect to the local vde socket \fIvdeplug\fP
instead pf stdin/stdout, using the libvdeplug library.
.TP
.B \-D 
Detach console and send the process to background.
.TP
.B \-b "\fIip.to.bind\fP"
The NS server will bind only to the specified ip, instead of any ip. 
Only valid in server mode (without -c)
.TP
.B \-c "\fIserver_addr\fP"
If specified, client mode is enabled. Vde_over_ns will try to send NS requests to \fIserver_addr\fP. If not specified, server mode is enabled by default.

.SH NOTICE
Virtual Distributed Ethernet is not related in any way with
www.vde.com ("Verband der Elektrotechnik, Elektronik und Informationstechnik"
i.e. the German "Association for Electrical, Electronic & Information
Technologies").

.SH SEE ALSO
\fBvde_switch\fP(1),
\fBdpipe\fP(1).
.br
.SH AUTHORS
VDE_OVER_NS is a tool by Daniele Lacamera <root@danielinux.net> 

VDE is a project by Renzo Davoli <renzo@cs.unibo.it>

based on nstx, a steganographic ns client/server software by:

	Florian Heinz <sky@sysv.de>

	Julien Oster <frodo@sysv.de>




Changelog
VDE 2.3.3, 2022-05-02
	* Several bug fixes
	* new hash function
	* Allow pipes on stdin for daemonized switches
	* fixed wirefilter '-N' option
	* vde_switch: min packet len enforced
	* Added vde_router
	* dpipe: Fixed error checking for setpgrp().
	* Fix incorrect chacha key size
	* configure: do not test for CXX, it's not needed
	* Fix description for "vlan/delport" command.
	* Silence compiler warning about implicit type.
	* Fix typo in error message
	* Fixes 2019.10.06
	* fstp: Add static to inline functions
	* configure: Check for wolfSSL instead of OpenSSL.
	* build: Fix parallel make.
	* vde_switch: add `-nostdin` flag to allow stdin to be closed
	* Deprecate modules
	* Fix warnings with newer gcc
	* [vde_plug] Removed kvde IPN support
	* In vde_switch retry send() when it returns ENOBUFS on Darwin and FreeBSD
	* configure.ac: Downgrade autoconf from 2.71 to 2.59
	* Added GitHub actions

VDE 2.3.2, 2011-11-23
	* dpipe: new daemon mode
	* [vde_switch] new mainloop management
	* [vde_switch] hash table based on 64bits integers
	* [vde_switch] new priority queue for packets
	* wirefilter: capacity renamed as "channel buffer size"

VDE 2.3.1, 2010-07-25
	* [vde_switch] Fix control socket path resolution regression.

VDE 2.3.0, 2010-05-29
	* Remove some hardcoded paths and use SYSCONFDIR/LOCALSTATEDIR instead.
	* [vde_switch] Fix multiple vlan tagging when packetfilter is used.
	* Build system fixes.
	* Ship README and patch to add VDE support in VirtualBox (included in
	  VirtualBox since 3.2.0).
	* Ship an updated version of slirp taken from QEMU and include VDE support.
	* [vde_cryptcab] Remove SIGALRM-based timers.
	* [vde_l3] Fix unicast test (sourceforge: #2726725)
	* [libvdeplug] Add vdestream abstraction to send/receive bytes (previously in
	  vde_plug)
	* [vde_switch] New command "port/createauto" to create a new port with an
	  automatically allocated ID

VDE 2.2.3, 2009-05-05
	* [vdeterm] Moved history management in a separate library.
	* [vde_switch] Console bug fixes.
	* Added vde_pcapplug (connect a vde_switch to a real interface using pcap)
	* Added missing includes of limits.h/ctype.h
	* [vde_switch] Added option --mgmtgroup to consmgmt module to specify the
	  group of the management socket.
	* [vde_switch] Added option --dirmode to datasock module to specify the mode
	  of the socket directory.
	* [vdetaplib] Added specification of port, mode and group.
	* [kvde_switch] Bugfixes.
	* [vde_switch] pdump plugin: add fifo support and buffered/unbuffered choice.
	* [common] Fix '#include <limits.h>' for PATH_MAX. (Closes: #2023790)
	* [vdetaplib] Manpage: added an example to override default port, group and
	  mode. (Closes: #2066885)
	* [vde_switch] Resize untagged ports bitarray while changing number of ports
	  (Thanks to Michael Sallaway). Closes: #2115547.
	* [vde_switch] Fix memory reset in BA_REALLOC and BAC_REALLOC (by rd235).
	  (Closes: #2123601)
	* [kvde_switch] By default kvde_switch is now not compiled in.
	* [vdeq] Fixed loop if an unknow parameter is specified. Thanks to tfero.
	  (Closes: #2134438)
	* [slirpvde] Imported patch from <ludwig.nussel@suse.de> to fix static buffer
	  overflow (Closes: #2138410).
	* [vde_switch] Fixed compatibility with armel architecture using "attribute
	  packet" on registration message structures (thanks to Michele Cucchi).
	* [vde_cryptcap] Some modifications to allow the client to work on the
	  OpenMoko freerunner.
	* [vde_switch] Added port access control.
	* [vde_switch] Changed regular user standard switch to ~/.vde2/stdsock.
	* [libvdeplug] Downsize vde_open_real() and vde_realpath() stack to fit in
	  User-Mode Linux.
	* [common] Add pkg-config files for VDE libraries.
	* [common] Add '--disable-pcap' configure option, thanks to Wulf C. Krueger
	  (Closes: #2600817).
	* [common] Fix bad malloc replacement (Closes: #2631581), thanks to Michele
	  Cucchi.
	* [vde_cryptcab] Added support for daemonization.
	* [vde_cryptcab] Fixed server side disconnections (thanks to Brendan Grieve).
	* [wirefilter] Added rc file.

VDE 2.2.2, 2008-07-08
	* Added --enable-profile to configure (for profiling)
	* Added cleanup() in vdeq on successfull termination (thanks huslu)
	* Added --mod to new syntax in vdeq man page
	* Fixed fd leak in runscript() (Closes: #2009311 on SF)
	* wirefilter man and code cleanup
	* wirefilter option for blinking added.
	* Gaussian distribution for range values, Gilbert bursty model for packet loss.
	* Bugfix on mgmt sessions.
	* Poor's man hub implemented. (patch by Luca Saiu)

VDE 2.2.1, 2008-06-17
	* New contact email address: <info@v2.cs.unibo.it>
	* Code cleanup (some more warnings removed)
	* [vde_crytcab] Cleared large memory leak due to missing
	  EVP_CIPHER_CTX_cleanup() after encryption/decryption
	* [vde_cryptcab] Cleared (very small) memory leak in crc32 calls
	* [vde_cryptcab] Fixed bug on handover after session timeout: now client
	  re-establishes the connection when it sees some traffic, after session
	  timeout.
	* Heavy directories reorganization, all source code moved under src and
	  include
	* New internal library libvdecommon.la with common utilities and LIBOBJ
	  replacement/compatibility functions
	* [common] Fix in poll emulation management
	* Fix to distclean problem due to wrong LIBOBJ management
	* [vde_switch] hash/setsize 0 was not recognized as an error and caused vde_switch
	  to hang due to an infinite loop
	* Don't compile cryptcab if configure detects not to.
	* Fix for autotools in Slackware
	* [vde_cryptcab] refactoring:
	  - simplified protocol (though keeping it back-compatible)
	  - fixed some bugs and memory leaks
	  - added some new features
	* [vde_switch] fix on BPDU generation/handling: do not generate BPDU if FSTP is not active
	  or we are HUB, forward BPDU if FSTP is not active (Closes #1943973 on SF).
	* [slirpvde] include limits.h on linux as well.
	* [vde_switch] Applied patch from Bjorn Mork to remove 32/64 bit confusion from bitarray.h

2.2.1/src/vde_over_ns/vde_over_ns.c
/* ----------------------------------------------------------------------------
 *
    VDE_OVER_NS 
	(C) 2007 Daniele Lacamera

    Derived from:
    NSTX -- tunneling network-packets over DNS

     (C) 2000 by Florian Heinz and Julien Oster

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License version 2, as 
    published by the Free Software Foundation.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

  -------------------------------------------------------------------------- */

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <fcntl.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <errno.h>
#include <time.h>
#include <stdlib.h>
#include <pwd.h>
#include <sysexits.h>
#include <syslog.h>

#include <config.h>
#include <vde.h>
#include <vdecommon.h>

#include "fun.h"
#include "pstack.h"
#include "dns.h"

#define DNSTIMEOUT 3
#define DRQLEN 10

#define BUFLEN 2000

static int nsid;

static void nstx_getpacket (void);
static struct nstx_senditem * alloc_senditem(void);
static void queue_senditem(const char *buf, int len);
static char *dequeue_senditem (int *len);
struct nstx_senditem * nstx_sendlist = NULL;
static char *vdesock = NULL;
static void nstxc_handle_reply(char *, int);
static int nstxc_send_packet(char *, int);

static void
usage(const char *prog, int code)
{
	fprintf (stderr, "usage: %s [-c DNSSERVER] [-s VDESOCK] [-i IP] [-D] <domainname>\n"
	    "Options:\n"
	    "\t-i IP (bind to port 53 on this IP only)\n"
	"\n"
	    "\t-D (call daemon(3) to detach from terminal)\n"
	"\n"
	    "\t-c DNSSERVER: Client mode. Tries to 'connect' to DNSSERVER.\n"
	    "\t\t(if this is not specified, server mode will be enabled by default.) \n"
	"\n"
	    "\t-s VDESOCKET: Attach to socket VDESOCKET \n"
	    "\t\t(if not specified, use stdin/stdout) \n"
	"\n"
	    "example:\n"
	    "\t%s -s /tmp/vde.ctl -c 1.2.3.4 tun.vdevirtualnetwork.foo [Client mode]\n"
	    "\t%s -s /var/vde-master-switch.ctl tun.vdevirtualnetwork.foo [Server mode]\n", prog, prog, prog);
	exit(code);
}

int main (int argc, char *argv[]) {
   signed char	 ch;
   const char *dir = NULL;
   in_addr_t	 bindto = INADDR_ANY;
   uid_t	 uid = 0;
   int		 daemonize = 0;
   int		 logmask = LOG_UPTO(LOG_INFO);
   char *clientmode_serveraddress=NULL;
   struct nstxmsg *msg;
   
   nsid = time(NULL);

   while ((ch = getopt(argc, argv, "Dh:i:s:c:")) != -1) {
	switch(ch) {
	case 'i':
		bindto = inet_addr(optarg);
		if (bindto == INADDR_NONE) {
			fprintf(stderr, "`%s' is not an IP-address\n",
			    optarg);
			exit(EX_USAGE);
		}
		break;
	case 'D':
		daemonize = 1;
		break;
	case 'g':
		logmask = LOG_UPTO(LOG_DEBUG);
		break;
	case 's':
		vdesock = strdup(optarg);	
		break;
	case 'h':
		usage(argv[0], 0);	/* no return */
	case 'c':
		clientmode_serveraddress = strdup(optarg);
		break;
	default:
		usage(argv[0], EX_USAGE);	/* no return */
	}
   }

   if (argc - optind < 1)
	usage(argv[0], EX_USAGE);

   dns_setsuffix(argv[optind]);
    
	
   if (uid && setuid(uid)) {
	syslog(LOG_ERR, "Can't setuid to %ld: %m", (long)uid);
	exit(EX_NOPERM);
   }
   if (daemonize && daemon(0, 0)) {
	syslog(LOG_ERR, "Can't become a daemon: %m");
	exit(EX_OSERR);
   }

   if (clientmode_serveraddress!=NULL){
	fprintf(stderr,"Client Mode\n");

	qsettimeout(10);
	open_ns(clientmode_serveraddress);
	init_vdesock(vdesock);

	for (;;) {
	    msg = nstx_select(1);
	    if (msg) {
	       if (msg->src == FROMNS) {
		  nstxc_handle_reply (msg->data, msg->len);
	       } else if (msg->src == FROMTUN) {
		  nstxc_send_packet (msg->data, msg->len);
	       }
	    }
	    timeoutqueue(NULL);
	    while (queuelen() < DRQLEN)
	      nstxc_send_packet (NULL, 0);
	  }

	  return 0;

   } else {
	fprintf(stderr,"Server Mode\n");
	  open_ns_bind(bindto);
	   init_vdesock(vdesock);
   
	if (dir) {
		/* Open the log-socket now (with LOG_NDELAY) before chroot-ing */
		openlog(argv[0], LOG_PERROR|LOG_PID|LOG_CONS|LOG_NDELAY, LOG_DAEMON);
		if (chroot(dir)) {
			syslog(LOG_ERR, "Can't chroot to %s: %m", dir);
			exit(EXIT_FAILURE); /* Too many possible causes */
		}
	   } else
		openlog(argv[0], LOG_PERROR|LOG_PID|LOG_CONS, LOG_DAEMON);

	   setlogmask(logmask);
		
		while (1)
		nstx_getpacket();
   
		exit(0);
   }
}

struct nstx_senditem * nstx_get_senditem(void) {
   struct nstx_senditem *ptr = nstx_sendlist;
   
   if (!nstx_sendlist)
     return NULL;
   
   ptr = nstx_sendlist;
   nstx_sendlist = nstx_sendlist->next;
   
   return ptr;
}

static void do_timeout (struct nstxqueue *q)
{
   struct dnspkt *pkt;
   int len;
   char *buf;
   
   pkt = dns_alloc();
   dns_setid(pkt, q->id);
   dns_settype(pkt, DNS_RESPONSE);
   dns_addanswer(pkt, "\xb4\x00\x00\x00", 4, dns_addquery(pkt, q->name));
   buf = (char*)dns_constructpacket (pkt, &len);
   sendns(buf, len, &q->peer);
   free(buf);
}  

void nstx_getpacket (void) {
   int len, link;
   const char *name, *buf, *data;
   struct nstxmsg *msg;
   struct nstxqueue *qitem;
   struct dnspkt *pkt;

   msg = nstx_select(1);
   
   if (msg) {
     if (msg->src == FROMNS) {
	pkt = dns_extractpkt((unsigned char*)msg->data, msg->len);
	if (pkt)
	  {
	     name = dns_getquerydata(pkt);
	     if (name)
	       {
		  syslog(LOG_DEBUG, "getpacket: asked for name `%s'",
			name);
		  queueitem(pkt->id, name, &msg->peer);
		  if ((data = dns_fqdn2data(name)) &&
		      (buf = nstx_decode((unsigned char*)data, &len)))
		    {
		       nstx_handlepacket(buf, len, &send_vde);
		    }
	       }
	     dns_free(pkt);
	  }
     } else if (msg->src == FROMTUN)
	  queue_senditem(msg->data, msg->len);
   }
   
   while (queuelen()) {
      if (!nstx_sendlist)
	break;
      qitem = dequeueitem(-1);
      pkt = dns_alloc();
      dns_setid(pkt, qitem->id);
      dns_settype(pkt, DNS_RESPONSE);
      link = dns_addquery(pkt, qitem->name);
      len = dns_getfreespace(pkt, DNS_RESPONSE);
      buf = dequeue_senditem(&len);
      dns_addanswer(pkt, buf, len, link);
      buf = (char*)dns_constructpacket(pkt, &len);
      sendns(buf, len, &qitem->peer);
   }
   timeoutqueue(do_timeout);
}



static struct nstx_senditem * alloc_senditem(void) {
   struct nstx_senditem *ptr = nstx_sendlist;

   if (!nstx_sendlist) {
      ptr = nstx_sendlist = malloc(sizeof(struct nstx_senditem));
   } else {
      while (ptr->next)
	ptr = ptr->next;
      ptr->next = malloc(sizeof(struct nstx_senditem));
      ptr = ptr->next;
   }

   memset(ptr, 0, sizeof(struct nstx_senditem));
   
   return ptr;
}

static void
queue_senditem(const char *buf, int len) {
   static int id = 0;
   struct nstx_senditem *item;
   
   item = alloc_senditem();
   item->data = malloc(len);
   memcpy(item->data, buf, len);
   item->len = len;
   item->id = ++id;
}

static char *
dequeue_senditem (int *len) {
   static char *buf;
   struct nstx_senditem *item = nstx_sendlist;
   struct nstxhdr *nh;
   int remain, dlen;
   
   remain = item->len - item->offset;
   dlen = *len - sizeof(struct nstxhdr);
   if (dlen > remain)
     dlen = remain;
   *len = dlen + sizeof(struct nstxhdr);
   buf = realloc(buf, *len);
   nh = (struct nstxhdr *)buf;
   memset(nh, 0, sizeof(struct nstxhdr));
   memcpy(buf+sizeof(struct nstxhdr), item->data + item->offset, dlen);
   nh->magic = NSTX_MAGIC;
   nh->seq = item->seq++;
   nh->id = item->id;
   item->offset += dlen;
   if (item->offset == item->len) {
      nh->flags = NSTX_LF;
      nstx_sendlist = item->next;
      free(item->data);
      free(item);
   }
   
   return buf;
}

static void nstxc_handle_reply (char * reply, int len) {
   struct dnspkt *pkt;
   const char *data;
   int datalen;
   
   pkt = dns_extractpkt ((unsigned char*)reply, len);
   if (!pkt)
     return;
   while ((data = dns_getanswerdata(pkt, &datalen))) {
      data = (char*)txt2data((unsigned char*)data, &datalen);
      nstx_handlepacket (data, datalen, &send_vde);
   }
   dequeueitem(pkt->id);
   dns_free(pkt);
}
  
static int nstxc_send_packet (char *data, int datalen) {
  static int id = -1;

  char *p;
  struct nstxhdr nh;
  struct dnspkt *pkt;
  int l;

  if (id < 0)
    id = time(NULL);
        
  nh.magic = NSTX_MAGIC;
  nh.seq = 0;
  nh.id = id++;
  nh.flags = 0;

  do {
    pkt = dns_alloc();
    dns_settype(pkt, DNS_QUERY);
    dns_setid(pkt, nsid);
    
    l = dns_getfreespace(pkt, DNS_QUERY);
    if (l <= 0) {
       printf("Fatal: no free space in dns-packet?!\n");
       exit(1);
    }
    p = malloc(l);
    l -= sizeof(nh);
    if (l > datalen) {
       l = datalen;
       nh.flags = NSTX_LF;
    }
    memcpy (p, (char*)&nh, sizeof(nh));
    if (data)
       memcpy (p + sizeof(nh), data, l);
    data += l;
    datalen -= l;
    
    dns_addquery(pkt, dns_data2fqdn(nstx_encode((unsigned char*)p, sizeof(nh)+l)));
    free(p);
    p = (char*)dns_constructpacket(pkt, &l);
    sendns(p, l, NULL);
    free(p);

    queueid(nsid);
    nsid++;
    nh.seq++;
  } while (datalen);

  return 0;
}
